python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +60,927 @@
+                        # caching this file.
+                        cache_msg = " and likely cached"
+
+                    msg = f"(via download to local file{cache_msg})"
+
+                log.debug(
+                    "Reading%s from location %s %s with formatter %s",
+                    f" component {log_component}" if log_component else "",
+                    uri,
+                    msg,
+                    self.name(),
+                )
+
+                with time_this(
+                    log,
+                    msg="Reading%s from location %s %s with formatter %s",
+                    args=(
+                        f" component {log_component}" if log_component else "",
+                        uri,
+                        msg,
+                        self.name(),
+                    ),
+                ):
+                    result = self.read_local_file(local_uri, component=component)
+
+                # File was read successfully so can move to cache
+                if can_be_cached:
+                    cache_manager.move_to_cache(local_uri, cache_ref)
+
+        return result
+
+    def read_cached_file(self, uri: ResourcePath, component: str | None = None) -> Any:
+        """Read a dataset from a, possibly cached, URI.
+
+        Parameters
+        ----------
+        uri : `lsst.resources.ResourcePath`
+            URI to use to read the dataset. This URI can be local or remote
+            and can refer to the actual resource or to a locally cached file.
+        component : `str` or `None`, optional
+            The component to be read from the dataset.
+
+        Returns
+        -------
+        in_memory_dataset : `typing.Any`
+            The Python object read from the resource.
+
+        Notes
+        -----
+        This method is only called if the class property
+        ``allow_remote_file_read`` is set to `True`.
+        """
+        raise NotImplementedError("This formatter does not know how to read a file.")
+
+    def read_local_file(self, local_uri: ResourcePath, component: str | None = None) -> Any:
+        """Read a dataset from a local file system.
+
+        Parameters
+        ----------
+        local_uri : `lsst.resources.ResourcePath`
+            URI to use to read the dataset. This URI is guaranteed to be
+            a local file.
+        component : `str` or `None`, optional
+            The component to be read from the dataset.
+
+        Returns
+        -------
+        in_memory_dataset : `typing.Any`
+            The Python object read from the resource.
+
+        Notes
+        -----
+        This method will only be called if `read_cached_file` was not called
+        or if `read_from_bytes` was called successfully.
+        """
+        raise NotImplementedError("This formatter does not know how to read a local file.")
+
+    def read_from_bytes(self, serialized_bytes: bytes, component: str | None = None) -> Any:
```

Comment:
I think a "read from file-like object" is more general than "read from bytes" and a complete replacement for it.  I wonder if it'd also be a replacement either of the other methods or something they could delegate to by default, but I haven't wrapped my head around the caching enough to have a sense for whether that's the case.